libp2p/NodeTable.cpp

```diff
@@ -90,46 +90,50 @@ void NodeTable::processEvents()
 
 bool NodeTable::addNode(Node const& _node, NodeRelation _relation)
 {
-    // TODO: Make private version of addNode() that could return reference to added node.
-    if (_relation == Known)
-    {
-        auto nodeEntry = make_shared<NodeEntry>(m_hostNodeID, _node.id, _node.endpoint);
-        // mark as validated
-        // TODO get last pong time as input, ping if needed
-        nodeEntry->lastPongReceivedTime = RLPXDatagramFace::secondsSinceEpoch();
-        DEV_GUARDED(x_nodes) { m_allNodes[_node.id] = nodeEntry; }
-        noteActiveNode(_node.id, _node.endpoint);
-        return true;
-    }
-
     if (!_node.endpoint || !_node.id)
     {
         LOG(m_logger) << "Supplied node has an invalid endpoint (" << _node.endpoint << ") or id ("
                       << _node.id << "). Skipping adding to node table.";
         return false;
     }
 
+    if (m_hostNodeID == _node.id)
+    {
+        LOG(m_logger) << "Skip adding self to node table (" << _node.id << ")";
+        return false;
+    }
+
+    bool bFound = false;
+    std::shared_ptr<NodeEntry> nodeEntry = nullptr;
     DEV_GUARDED(x_nodes)
     {
-        auto const it = m_allNodes.find(_node.id);
-        if (it != m_allNodes.end())
-        {
-            LOG(m_logger) << "Node " << _node.id << "@" << _node.endpoint
-                          << " is already in the node table";
-            return true;
-        }
+        auto nodePair = m_allNodes.insert(
+            {_node.id, make_shared<NodeEntry>(m_hostNodeID, _node.id, _node.endpoint)});
```

Comment:
This way you create a shared_ptr unnecessary when an entry is already in the table. That's you I asked to insert nullptr first which later can be replaced with correct shared_ptr if needed.